Adding utils and tests for hello.py

monchier opened the pull request:
The purpose of this PR is adding a test for some utility in `hello.py`. I created a package for `hello.py`, modified `cli.py` and I added a util file to be used by `hello.py`. I then added a test for such util file.

monchier commented on lib/streamlit/hello/hello.py:
```diff
@@ -17,20 +17,24 @@
+import streamlit as st
+from streamlit.logger import get_logger
+
+from . import util
```
Python2.7 ok?

tvst replied:
This won't work with py2. But I think you can import `absolute_import` from `__future__` to make it work.

monchier commented on lib/streamlit/hello/util.py:
```diff
@@ -0,0 +1,62 @@
+    in conjunction of inspect.getsourcelines(...) to get the source code
+    lines of a function. If the docstring is longer than MAX_DOCSTRING lines
+    it will not remove the docstring and raise an exception. The function also
+    assumes that the input source code is syntactically correct.
```
I wonder we should add a check, like, trying to compile the code. In a way, this is meant to be used by hello.py, but some guarantee against future uses is a good thing. I wonder this is too heavy and the user of this function should guarantee correctness.

monchier commented on lib/streamlit/hello/util.py:
```diff
@@ -0,0 +1,62 @@
+    while '"""' not in lines[index]:
+        index += 1
+        # limit to MAX_DOCSTRING lines, if the docstring is longer, just bail
+        if index > MAX_DOCSTRING:
```
making sure we don't stay in this loop forever. It is the only correctness check I have added. If the docstring is syntactically correct this is not needed.

monchier commented on lib/streamlit/hello/util.py:
```diff
@@ -0,0 +1,62 @@
+        a copy of lines where the docstring is removed
+    """
+    if len(lines) == 0:
+        raise Exception("You should pass code with a function declaration included")
```
Should I make my own `Exception`?

tvst commented on lib/streamlit/hello/util.py:
```diff
@@ -0,0 +1,62 @@
+
+def remove_declaration_and_docstring(lines):
+    """
+    This function parses the source code of a function and removes the function
```
Use Numpy doc style. The first sentence must be a summary that fits in one line. And shouldn't say "This function" or use present tense "parses". Instead just use the imperative "parse".

Example:
> """Return a function's source code with the docstring removed.
> 
> This function parses the source code (...)
> """

tvst commented on lib/streamlit/hello/util.py:
```diff
@@ -0,0 +1,62 @@
+    Returns
+    -------
+    [str]
+        a copy of lines where the docstring is removed
```
Numpy doc requires a blank line here

tvst commented on lib/streamlit/hello/util.py:
```diff
@@ -0,0 +1,62 @@
+    """
+    if len(lines) == 0:
+        raise Exception("You should pass code with a function declaration included")
+    # lines contains only the function declaration
```
Capitalize comments

monchier replied:
It is a variable name in this case.

tvst commented on lib/streamlit/hello/__init__.py:
```diff
@@ -0,0 +1,18 @@
+import streamlit.hello
+
+if __name__ == "__main__":
+    streamlit.hello.run()
```
Why is this needed? We could just have `cli.py` call `run()`, right?

monchier replied:
Yeah,... +1

monchier replied:
I did ended up removing this and refactoring the imports.

tvst commented on lib/streamlit/hello/util.py:
```diff
@@ -0,0 +1,62 @@
+
+    Parameters
+    ----------
+    lines : [str]
```
Use Numpy doc style

tvst commented on lib/streamlit/hello/util.py:
```diff
@@ -0,0 +1,62 @@
+
+    Returns
+    -------
+    [str]
```
Use Numpy doc style

tvst commented on lib/streamlit/hello/util.py:
```diff
@@ -0,0 +1,62 @@
+MAX_DOCSTRING = 1000
+
+
+def remove_declaration_and_docstring(lines):
```
You can get a function's docstring with `the_func.__doc__`. Wouldn't it be easier / safer to just use that here?

For example:
```
doc = the_func.__doc__.splitlines()
last_line = len(doc)
return lines[last_line + 1:]
```

monchier replied:
I can try this. Seems a good option. Not sure on top of my head if `the_func.__doc__.splitlines()` has to maintain the same formatting and gives the same number of lines, but it may as well be.

monchier replied:
It does not tell me about the lines with `"""`, it only gives me the lines with the actual content of the docstring. So,
```
"""
"""
```
and 
```
"""  """
```
are the same. I would still have to parse and do something with these, so not sure of the actual advantage.

monchier replied:
I tried to write it and it quickly became as complex. I would leave it as it is. I am not a big fan of ad hoc logic like this one...

monchier replied:
Honestly there might not be need for doing all this introspection... we could just remove the docstring and factor out the comments, then require all demo functions *not* to have a docstring (easier assert). The current code happened because of some initial back and forth between me and Adrian, but no reason to stay this way. If you like this approach better I can refactor in this sense, it will take an extra hour or so.

monchier commented:
I removed the `remove_docstring` function and removed the use of introspection to get the description of the functions via docstring. Now, the description is passed as an item in the map that describes the demos (`DEMOS` in `hello.py`)

I also moved the actual demo code out to demos.
